### ISSUE TITLE
Profile reordering for Settings UI

### ISSUE THREAD

**Issue by carlos-zamora** [microsoft/terminal#8914]
# Description of the new feature/enhancement

The Settings UI doesn't have a way of reordering your profiles. In the JSON, we're able to just reorder the array of `profiles.list`.

Some considerations to keep in mind include:
- [**New Tab Menu Customization**](https://github.com/microsoft/terminal/blob/main/doc/specs/%231571%20-%20New%20Tab%20Menu%20Customization/%231571%20-%20New%20Tab%20Menu%20Customization.md):
  - allow actions in the dropdown
  - nesting profiles/actions in the dropdown

_Ideally_, we could have a design that's future-proof. But since we don't know when the above will be implemented, we could just go with a simple non-future-proof design now, then redesign later.

# Proposed technical implementation details (optional)

A few ideas have come to mind, but @cinnamon-msft and I haven't actually taken the time to sit down and design this yet. I'll list a few ideas here, but I'm excited to see what ideas the community can come up with:

## 1. Drag and drop NavigationView's menu items

Pretty straightforward. The idea is that you can drag and drop the menu items for profiles.

![NavigationView](https://user-images.githubusercontent.com/11050425/106053841-1d725c80-60a0-11eb-95ac-758887b344cb.png)

Concerns:
- knowing that those menu items are drag-able isn't intuitive. Maybe we could add some kind of 
gripper icon? 
![gripper icon example](https://user-images.githubusercontent.com/11050425/106054056-5f9b9e00-60a0-11eb-8c39-152d886709db.png)
- not really future-proof

## 2. Add an index "setting" in Profile > General

This would probably be a number box on each Profile > General page (not base layer).

Concerns:
- don't like that you can't see all the overall order

**Comment by carlos-zamora** [comment]
@cinnamon-msft assigning you for now since we need a design. I don't think we'll need to write up a design doc though, tbh, but we'll play it by ear.

**Comment by ultrasound1372** [comment]
I think if this were implemented keyboard accessibility would also need to be considered. I was personally thinking the up/down buttons like present in the path environment variable dialog in system properties, but not sure if that would visually mesh well with this UI. If it uses dragging I'd think something like alt+up/down could work if we're only moving one profile at a time. Would reordering be the only way for me to change the control+shift+number layout to activate the 11th profile in my list? Said 11th profile happens to be a VS native tools prompt and I'd like that to go above all my MSYS prompts.

**Comment by BuzyBeeC** [comment]
Perhaps a hybrid solution, where profiles are displayed as a numbered list (indexed) in the settings, would combine the best of each. On the left of each profile icon would a be a number “text box” that can be changed via typing. The GUI could visibly reorder profiles based on this number to make it clear. It would intuitively fit with the ctrl+shift+<0-9> shortcut. It could also probably be made keyboard accessible as @ultrasound1372 mentioned but much more efficient than moving one position at a time. Since the GUI would already update position in the list, a reverse drag-and-drop option that updates the index could maybe be added as a bonus (with gripper icon on the right). I don’t it would be too unintuitive since many rearranging actions in modern GUIs are drag-and-drop. After all, I first stumbled upon this thread after trying to drag to reorder the profiles.

**Comment by ultrasound1372** [comment]
If @BuzyBeeC's thing is considered, I think that that numeric edit should be a spin control rather than a simple edit. That would also allow range limits to be placed and cut down on input validation requirements. You can't change this profile to index 20 when you only have 10 profiles, after all. And for keyboard users this would let us change the ordering with simple up/down arrow on the box.

**Comment by BuzyBeeC** [comment]
> If @BuzyBeeC's thing is considered, I think that that numeric edit should be a spin control rather than a simple edit. That would also allow range limits to be placed and cut down on input validation requirements. You can't change this profile to index 20 when you only have 10 profiles, after all. And for keyboard users this would let us change the ordering with simple up/down arrow on the box.

This sounds like a simpler and more user-friendly approach.

**Comment by brian-kneebone-hp-com** [comment]
I'm more of a fan of click/drag (for mouse/finger users) and the keyboard options proposed above for those that require it (I'd assume CTRL+UP/DOWN or ALT+ UP/DOWN or whatever convention I'm not aware of that makes the most sense would be nice). Why create a new UI if it's not absolutely required. At least for me, my first instinct was to try these techniques, but I might be a weirdo.

**Comment by ultrasound1372** [comment]
Unfortunately, there is no standard for keyboard only reordering or drag and drop. This would be one of the first applications I know of to implement such a thing, second in my knowledge only to foobar2000's list of decoders. Everything else is either use cut/paste or you're just dead out of luck if you can't see. Depending on if you're going to allow multiple movements at once I would go with either alt or control as they said. If you can drag multiple consecutive profiles at once, to reorder a group, I'd select those with shift and move them with alt. Otherwise I'd use control. If you let people select with shift and use control someone might get confused with the other thing where you can control arrows to avoid changing the selection and space to toggle it for the currently focused item, allowing noncontiguous selection. For reordering though that doesn't make sense.

**Comment by psztoch** [comment]
Profile reorder is important feature. Should be added ASAP.

**Comment by zadjii-msft** [comment]
@psztoch You can always reorder the profiles manually in the `settings.json` file before this gets added.

**Comment by psztoch** [comment]
I do this to myself. But regular users (programmers) complain and don't want to use powershell 7 which is added at the end.

**Comment by zs-dima** [comment]
+1

**Comment by ggjulio** [comment]
@zadjii-msft But unfortunately it doesn't change profile's shortcuts. Which I think it's the most important feature of this request. Because I think we don't cares much about the order of the profiles shown in `settings`, we set them one time and then forget about the `settings` page, whereas shortcuts... :-) (I would love to have the first `ubuntu` as  `CTRL+SHIFT+1` instead of no shortcuts)

![image](https://user-images.githubusercontent.com/13398285/214956539-35001468-e480-4382-a68f-986eb0e52136.png)

**Comment by zadjii-msft** [comment]
Uh, it absolutely does....?

![image](https://user-images.githubusercontent.com/18356694/214968367-7aef7658-d89c-4f74-8c65-2e6f43e66f0d.png)

Also, feel free to just rebind the actions to whatever you want:

```json
        { "command": { "action": "newTab", "index": 0 }, "keys": "ctrl+shift+1" },
        { "command": { "action": "newTab", "index": 1 }, "keys": "ctrl+shift+2" },
        { "command": { "action": "newTab", "index": 2 }, "keys": "ctrl+shift+3" },
        { "command": { "action": "newTab", "index": 3 }, "keys": "ctrl+shift+4" },
        { "command": { "action": "newTab", "index": 4 }, "keys": "ctrl+shift+5" },
        { "command": { "action": "newTab", "index": 5 }, "keys": "ctrl+shift+6" },
        { "command": { "action": "newTab", "index": 6 }, "keys": "ctrl+shift+7" },
        { "command": { "action": "newTab", "index": 7 }, "keys": "ctrl+shift+8" },
        { "command": { "action": "newTab", "index": 8 }, "keys": "ctrl+shift+9" },

```

Just change the `index` to whichever `index` profile you want.

**Comment by ggjulio** [comment]
That's weird it doesn't change shortcuts list for me. (terminal version: `1.15.3465.0`)
![image](https://user-images.githubusercontent.com/13398285/215044178-404288b9-b533-412c-a388-40ff84825a86.png)

Thanks for showing me the rebind workaround. :-)
It works great however it doesn't update the hint's list:
![image](https://user-images.githubusercontent.com/13398285/215047546-d9919615-fdc9-43b7-8290-ccbabe33bb63.png)
Anyway, that's not a big deal.

Thanks @zadjii-msft

**Comment by zadjii-msft** [comment]
@ggjulio FWIW, the first profile you have in your `list` is not actually your Ubuntu distro. That one's marked `hidden: true`. I'd reckon there's another Ubuntu profile, probably with a `source` that mentions "Canonical", right below the azure cloud shell 😉

**Comment by ggjulio** [comment]
Ho crap, I'm dumb, didn't see this this true property...sorry for all the spam,
I'll delete all those useless messages...
Thank for your time

**Comment by zadjii-msft** [comment]
Don't worry about it! Sometimes, things like that can be helpful for other folks to read and learn from too ☺️

**Comment by tietomattias** [comment]
@ggjulio I had the same mistake in my JSON >.<

**Comment by NewtonChutney** [comment]
Ahh.. Was just about to raise a duplicate issue! 😅
+1 on this request
BTW, is the backlog prioritized based on the 👍 hits?

**Comment by zadjii-msft** [comment]
> BTW, is the backlog prioritized based on the 👍 hits?

More or less, yea.

**Comment by ReDNeQ** [comment]
That Drag and Drop tho! Sure I edited the JSON file, but man o man, some of us want it all. ;)

**Comment by raghuwanshi** [comment]
> # You can always reorder the profiles manually in the `settings.json` file before this gets added.

We used to customize everything manually in the `settings.json`, then why do we even have a `Settings` GUI? The answer is simple 'ease of use' or user friendliness.

**Comment by eduarddejong** [comment]
> > # You can always reorder the profiles manually in the `settings.json` file before this gets added.
> 
> We used to customize everything manually in the `settings.json`, then why do we even have a `Settings` GUI? The answer is simple 'ease of use' or user friendliness.

The reason I appreciate the help of a GUI for the Terminal is that the settings JSON file is still not as easy to restore from a backup version as this is the case with the one of VS Code for example. And I mean like after a Windows reinstall or on another machine.
With a bit of file comparing and merging using VS Code or Meld we'll get the job done, but there are always some profiles with unique GUIDs, etc., that need to be corrected on the new system.
And I also don't know about any automated syncing solution yet.
So in other words, we simply cannot argu that that JSON file editing "should" be the preferred or even intended way of settings editing.

**Comment by Agent-E11** [comment]
I would like to say that this feature would be greatly appreciated

**Comment by sedlund** [comment]
> I would like to say that this feature would be greatly appreciated

Everyone here has been waiting for you to say this.

**Comment by allusernamestakenexceptthis** [comment]
Based on no.2 solution.  Why not add sort_order key/value that accept any integer and in settings of profile as numeric textbox.  default is 10 with increment of 10. e.g. 10,20,30,40,50.  that way it's easy for users to just change the one they want or set profile to be always last like 9999.

The menu then reorders based on this Ascending.  For most users that needs it, this would do the job. and should be easy to implement.  Perhaps can be added to advanced since not a very friendly design.

**Comment by Nikhil-Software-Cartel** [comment]
https://github.com/microsoft/terminal/assets/54004431/df6f3101-a470-4ed8-8ed2-947a991b8717

@allusernamestakenexceptthis @eduarddejong @raghuwanshi @ReDNeQ @zadjii-msft @NewtonChutney @tietomattias @ggjulio @ultrasound1372 @psztoch @brian-kneebone-hp-com @BuzyBeeC @carlos-zamora 

**please have a look at the video.** and code https://github.com/Nikhil-Software-Cartel/TerminalProfileManager (make sure to view last section)🙏🥺.
Context, wrong order was killing productivity for me. so spent whole day for this. it may have some rough edges, but kind of works, atleast on my pc. pls try and tell.
<details><summary>Features</summary>
<p>

- drag and drop profiles to reorder them, and save the new order to the settings file.
- intelligently read settings from `AppData/Local` folder irrespective of user name.
- color theme matched to windows terminal dark mode.
- edit/extend logic using simple python code and change GUI using html and css.
- if python already installed, half the setup is done.
- jquery, jquery-UI and bootstrap used for GUI. Keyboard support in future.
- hidden profiles explicitly dimmed.

</p>
</details>

**Comment by Nikhil-Software-Cartel** [comment]
### Terminal direct integration is now added!
@carlos-zamora 
![image](https://github.com/microsoft/terminal/assets/54004431/f3d77fa8-3b60-4021-a791-be4ed1394197)

**Comment by NewtonChutney** [comment]
> https://github.com/microsoft/terminal/assets/54004431/df6f3101-a470-4ed8-8ed2-947a991b8717

@Nikhil-Software-Cartel this looks more like a PoC but I think the idea here is to have the list in the main terminal settings page itself be draggable

**Comment by NewtonChutney** [comment]
> **please have a look at the video.** and code https://github.com/Nikhil-Software-Cartel/TerminalProfileManager (Important: make sure to view last section of repo)🙏🥺.
> 
> Context, wrong order was reducing productivity for me. so built this. but kind of works, atleast on my pc. pls try and tell. 
> @allusernamestakenexceptthis @eduarddejong @raghuwanshi @ReDNeQ @zadjii-msft @NewtonChutney @tietomattias @ggjulio @ultrasound1372 @psztoch @brian-kneebone-hp-com @BuzyBeeC @carlos-zamora 

And please don't degrade yourself to be considered spam or a nuisance by forcing things on others.. 
Just post your solution or work.. Interested folks will revert

**Comment by trparky** [comment]
When will this be something that's added to Terminal?

**Comment by Agent-E11** [comment]
@trparky It looks like it is planned https://github.com/microsoft/terminal/issues/10000

**Comment by Agent-E11** [comment]
@caihongxu It is way faster to open the `settings.json` file and reorder them there. The list is located at `profiles.list`

**Comment by caihongxu** [comment]
Ah missed that part. That's a great idea thanks.

**Comment by FaffeF** [comment]
I find it very confusing that there is now a "New Tab Menu" section in the Settings, but the order of the ctrl-shift-# shortcuts is not affected by that configuration. Instead those shortcuts are decided based on the more or less random order in which the profiles have been created. Perhaps if the profile numbering was exposed in the "New Tab Menu" section, actual profile order would not matter anymore?

**Comment by carlos-zamora** [comment]
> I find it very confusing that there is now a "New Tab Menu" section in the Settings, but the order of the ctrl-shift-# shortcuts is not affected by that configuration.

Honestly, that's probably the last thing we should fix before closing this issue out, right? The New Tab Menu page allows for profile reordering already as well as adding folders, separators, and profile matchers.

> Instead those shortcuts are decided based on the more or less random order in which the profiles have been created.

Pretty much. The way it works is that the `"index"` parameter for the `"newTab"` action targets the profile in the profile list in the JSON. This worked great at first, but once new tab menu customization was added, this got left behind and now the index doesn't match up with the order in the new tab menu.

An open question that I'd like feedback on is, how is the indexing expected to work with folders, a profile matcher entry, and the remaining profiles entry? Consider the following scenarios for a new tab menu:

Remaining Profiles & Profile Matcher:
```
┟ Profile A
┟ Profile B
┖ Remaining Profiles (or Profile Matcher): assume profiles C, D, E are added
```

Folders:
```
┟ Profile A
┟ Profile B
┟ Folder
┃┟ Profile C
┃┖ Profile D
┟ Profile E
┖ Profile F
```

Combination of both:
```
┟ Profile A
┟ Profile B
┟ Folder
┃┖ Profile Matcher for WSL
┟ Folder
┃┖ Remaining Profiles
┖ Profile C
```

For these scenarios, consider these questions:
- What does each value for `newTab.index` do?
   - 1 --> Profile A; 2 --> Profile B; But should 3 --> open the folder? or open the first profile in the folder? Or target the first profile not in a folder?
- What happens when profiles are dynamically added/removed by the profile matcher or remaining profiles?
- What if `folder.inline` is enabled (the folder is replaced with a single profile when there is only one profile in the folder)?

> Perhaps if the profile numbering was exposed in the "New Tab Menu" section, actual profile order would not matter anymore?

Oooh. I like that. It makes it less ambiguous. Maybe the way we can display that in the settings UI is just like it's shown in the dropdown? Add in a "_Ctrl+Shift+#_" in the list view to make it clear what keybinding is associated.

**Comment by FaffeF** [comment]
I think the first case is clear cut: A1 B2 C3 D4 E5

If the profiles in the folder case are all manually added I would default to the order they appear in the New Tab Menu, i.e. the profiles in the folder are invisible. Tack them to the end if any numbers are left: A1 B2 E3 F4 C5 D6

Last case, "remaining" always go at the end: A1 B2 C3 WSL[4-x] Remaining

Though I would much prefer if this is just the default: expose the shortcuts in the New Tab Menu and let us decide. I have ctrl-shift-4 in muscle memory for WSL ubuntu, and Kali at ctrl-shift-5. If I installed another computer with only Kali, I would prefer it could be set to 5 whether Ubuntu is installed or not.

**Comment by DHowett** [comment]
Begone, spam bot

**Comment by DHowett** [comment]
> If I installed another computer with only Kali, I would prefer it could be set to 5 whether Ubuntu is installed or not.

So, you can. It may not reflect properly in the UI, but if you add this to `settings.json` in the `actions` section:

```json
        {
            "command": 
            {
                "action": "newTab",
                "profile": "Kali"
            },
            "keys": "ctrl+shift+5"
        }
```

it will override the built-in binding for <kbd>ctrl+shift+5</kbd> and always launch Kali.

**Comment by FaffeF** [comment]
Sure, but this was in the context of the settings GUI. I could do all of this in notepad years ago I guess, but it's not convenient.

**Comment by DHowett** [comment]
Yeah, that's fair. Often I'll take a minor inconvenience over waiting even an hour for something, so I tend to lean more towards "get done now, find the right way to do it later." 🙂

**Comment by nikhil-swamix** [comment]
whats the ETA for > Profile reordering for Settings UI ?

**Comment by longebane** [comment]
Were we leaning towards drag and drop for the GUI? (in the Profiles submenu). I read the comments here and it wasn't clear to me what will be the approach

**Comment by ultrasound1372** [comment]
As long as there's an accessible alternative

**Comment by phntsm** [comment]
I would very much like a drag-drop feature as well. On occasion I've avoided adding new profiles just because I didn't feel like reorganizing them in the settings file at the time hahaha 😅
